File: PheromoneGrid.py
import numpy as np
from collections.abc import Iterator

import Random
from EnvironmentData import EnvironmentData
from Pheromone import Pheromone
import Optim


class PheromoneGrid:
    __env_data: EnvironmentData
    __grid: np.ndarray

    def __init__(self, width: int, height: int, data: EnvironmentData):
        self.__env_data = data

        self.update_size()

        for i in range(self.__env_data.random_natural_pheromones_count):
            x = int(Random.get_random_float() * width)
            y = int(Random.get_random_float() * height)

            self.__grid[x, y].deactivate()
            self.__grid[x, y].level = self.__env_data.natural_pheromones_strength

    def __getitem__(self, pos: tuple[int, int]) -> Pheromone:
        return self.__grid[pos]

    def __iter__(self) -> Iterator[Pheromone]:
        return self.__grid.__iter__()

    def update(self) -> None:
        for p_a in self.__grid:
            for p in p_a:
                if p.get_active() and p.get_level() > 0:
                    p.diffuse()
                    # Optim.bank_exec("Diffuse", p.diffuse)
        # print(Optim.time_bank)
        # Optim.reset_bank()

    def add_slime_level(self, x: int, y: int) -> None:
        self.__grid[x, y].add_slime_level()

    def set_slime_id(self, x: int, y: int, p_id: int) -> None:
        self.__grid[x, y].set_slime_id(p_id)

    def get_grid(self) -> np.ndarray:
        return self.__grid

    def update_size(self) -> None:
        self.__grid = np.ndarray((self.__env_data.grid_width, self.__env_data.grid_height), Pheromone)

        for x in range(self.__env_data.grid_width):
            for y in range(self.__env_data.grid_height):
                self.__grid[x, y] = Pheromone(self.__env_data, x, y)

File: PheromoneGrid.cpp
//
// Created by zwazo on 2022-09-24.
//

#include "PheromoneGrid.h"
#include "Random.h"

#include <iostream>

PheromoneGrid::PheromoneGrid(int width, int height, EnvironmentData *data) {
    envData = data;

    updateSize();

    for (int i = 0; i < envData->random_natural_pheromones_count; i++) {
        int x = (int) (Random::getRandomDouble() * width);
        int y = (int) (Random::getRandomDouble() * height);

        grid->at(y).at(x)->deactivate();
        grid->at(y).at(x)->level = envData->natural_pheromones_strength;
    }
}

void PheromoneGrid::update() {
    for (std::vector<Pheromone*> &line: *grid) {
        for (Pheromone *pheromone: line) {
            pheromone->diffuse();
        }
    }
}

void PheromoneGrid::addSlimeLevel(int x, int y) {
    grid->at(y).at(x)->addSlimeLevel();
}

void PheromoneGrid::setSlimeId(int x, int y, long id) {
    grid->at(y).at(x)->slimeId = id;
}

std::vector<std::vector<Pheromone*>> *PheromoneGrid::getGrid() {
    return grid;
}

void PheromoneGrid::updateSize() {
    grid = new std::vector<std::vector<Pheromone*>>();

    for (int y = 0; y < envData->grid_height; y++) {
        std::vector<Pheromone*> v = std::vector<Pheromone*>();

        for (int x = 0; x < envData->grid_width; x++) {
            v.push_back(new Pheromone(envData));
        }
        grid->push_back(v);
    }
}
